datastore/src/main/java/io/spine/server/storage/datastore/query/DsFilters.java

```diff
@@ -0,0 +1,184 @@
+     * {@code Collection} of query parameters. The difference is in
+     * <ul>
+     *     <li>type: from a {@code QueryPredicate} to Datastore {@code Filter};
+     *     <li>format: all the resulting {@code Filter}s are conjunctive (as Datastore does not
+     *         support disjunction) and are related with each other in the disjunctive way.
+     * </ul>
+     *
+     * <p><i>Example:</i>
+     *
+     * <p>Given query predicates {@code p1}, {@code p2}, {@code p3}, {@code p4}, {@code p5} passed
+     * into the method within the following construction: {@code p1 & (p2 | p3) & (p4 | p5)}. Then
+     * the resulting {@code Collection} of {@code Filters} will be constructed as
+     * {@code (p1 & p2 & p4) | (p1 & p2 & p5) | (p1 & p3 & p4) | (p1 & p3 & p5)}.
+     *
+     * <p>The separate conjunctive groups (e.g. {@code (p1 & p2 & p4)}) in the result
+     * {@code Collection} are placed into a single {@code Filter} instances one per group.
+     *
+     * <p>In other words, the predicate expression is brought into the
+     * <a href="https://en.wikipedia.org/wiki/Disjunctive_normal_form">disjunctive normal form</a>.
+     *
+     * <p>Note that by the convention, the distinct {@code QueryPredicate} instances
+     * are considered to be joined by the {@code &} operator. However, the resulting {@code Filter}
+     * instances are specified to be joined with {@code |} operator, so that the merged result
+     * of executing the queries with each of the filters will be the result of the whole expression.
+     *
+     * <p>If the given parameter {@code Collection} is empty, and empty {@code Collection}
+     * is returned.
+     *
+     * @param predicate
+     *         the predicate to convert
+     * @param adapter
+     *         an adapter performing the required type conversions
+     * @return the equivalent expression of in Datastore {@code Filter} instances
+     */
+    static <R extends Message> Collection<StructuredQuery.Filter>
+    fromPredicate(QueryPredicate<R> predicate, FilterAdapter adapter) {
+        checkNotNull(predicate);
+        checkNotNull(adapter);
+
+        Collection<StructuredQuery.Filter> results;
+        if(predicate.isEmpty()) {
+            results = emptySet();
+        } else {
+            results = toDsFilters(predicate, adapter);
+        }
+        return results;
+    }
+
+    private static <R extends Message> Collection<StructuredQuery.Filter>
+    toDsFilters(QueryPredicate<R> predicate, FilterAdapter adapter) {
+        ImmutableList.Builder<StructuredQuery.Filter> result = ImmutableList.builder();
+        QueryPredicate<R> dnf = predicate.toDnf();
+        ColumnMapping<Value<?>> mapping = adapter.columnMapping();
+        if (dnf.operator() == AND) {
+            StructuredQuery.Filter group = handleConjunctiveGroup(dnf, mapping);
+            result.add(group);
+        } else {
+            checkState(dnf.allParams()
```

Comment:
Please extract this check into a method. It would 1) help to eliminate the funny wrapping, and 2) please consider documenting the reason for the check.

Reply:
You have helped me spotting a nasty bug!

Fixed now.

Thank you!